src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ParameterCapturingPipeline.cs

```diff
@@ -0,0 +1,195 @@
+
+using Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing.Eventing;
+using Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing.FunctionProbes;
+using Microsoft.Diagnostics.Monitoring.StartupHook.MonitorMessageDispatcher.Models;
+using Microsoft.Diagnostics.Tools.Monitor.ParameterCapturing;
+using Microsoft.Diagnostics.Tools.Monitor.Profiler;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing
+{
+    internal sealed class ParameterCapturingPipeline : IDisposable
+    {
+        private sealed class CapturingRequest
+        {
+            public CapturingRequest(StartCapturingParametersPayload payload)
+            {
+                Payload = payload ?? throw new ArgumentNullException(nameof(payload));
+                StopRequest = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+            }
+
+            public StartCapturingParametersPayload Payload { get; }
+            public TaskCompletionSource StopRequest { get; }
+        }
+
+        private readonly IFunctionProbesManager _probeManager;
+        private readonly ILogger _logger;
+        private readonly Channel<CapturingRequest> _requestQueue;
+        private readonly ConcurrentDictionary<Guid, CapturingRequest> _allRequests = new();
+
+        public ParameterCapturingPipeline(ILogger logger, IFunctionProbesManager probeManager)
+        {
+            _logger = logger;
+            _probeManager = probeManager;
+
+            _requestQueue = Channel.CreateBounded<CapturingRequest>(new BoundedChannelOptions(capacity: 1)
+            {
+                FullMode = BoundedChannelFullMode.DropWrite
+            });
+        }
+
+        public async Task RunAsync(CancellationToken stoppingToken)
+        {
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                CapturingRequest request = await _requestQueue.Reader.ReadAsync(stoppingToken);
+                if (!TryStartCapturing(request.Payload))
+                {
+                    continue;
+                }
+
+                using CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken);
+                cts.CancelAfter(request.Payload.Duration);
+
+                try
+                {
+                    await request.StopRequest.Task.WaitAsync(cts.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+
+                }
+
+                _logger.LogInformation(ParameterCapturingStrings.StopParameterCapturing);
+                _probeManager.StopCapturing();
+                ParameterCapturingEventSource.Log.CapturingStop(request.Payload.RequestId);
```

Comment:
Rather than accessing a static instance, create a callback interface that the ParameterCapturingService (or an inner class) can implement and effectively forwards to the event source. This would also better separate this implementation so that it can be tested without automatically involving the event source and could allow tests to verify state of the pipeline by implementing its own callback. I believe this would also alleviate the need for the event source to be a static instance, and thus can be a member instance that is actually disposed properly.